engine/src/conversion/analysis/fun/implicit_constructors.rs

```diff
@@ -59,254 +75,441 @@ struct ExplicitFound {
-                })
-            });
-            // Conservatively, we will not generate implicit constructors for any struct/class
-            // where we don't fully understand all field types. We need to extend our knowledge
-            // to understand the constructor behavior of things in known_types.rs, then we'll
-            // be able to cope with types which contain strings, unique_ptrs etc.
-            let any_field_or_base_not_understood = bases
-                .iter()
-                .chain(field_types.iter())
-                .any(|qn| !all_known_types.contains(qn));
-            let explicit_items_found = ExplicitItemsFound {
-                move_constructor: find(ExplicitKind::MoveConstructor),
-                copy_constructor: find(ExplicitKind::ConstCopyConstructor)
-                    || find(ExplicitKind::NonConstCopyConstructor)
-                    || find(ExplicitKind::DeletedOrInaccessibleCopyConstructor),
-                any_other_constructor: find(ExplicitKind::OtherConstructor),
-                any_bases_or_fields_lack_const_copy_constructors,
-                any_bases_or_fields_have_deleted_or_inaccessible_copy_constructors,
-                any_bases_have_deleted_or_inaccessible_destructors,
-                destructor: find(ExplicitKind::Destructor)
-                    || find(ExplicitKind::DeletedOrInaccessibleDestructor),
-                copy_assignment_operator: find(ExplicitKind::CopyAssignmentOperator),
-                move_assignment_operator: find(ExplicitKind::MoveAssignmentOperator),
-                has_rvalue_reference_fields: details.has_rvalue_reference_fields,
-                any_field_or_base_not_understood,
+            let get_items_found = |qn: &QualifiedName| -> Option<ItemsFound> {
+                if known_types().is_known(qn) {
+                    Some(known_type_items_found(qn))
+                } else {
+                    all_items_found.get(qn).copied()
+                }
             };
-            log::info!(
-                "Explicit items found for {:?}: {:?}",
-                name,
-                explicit_items_found
-            );
-            let implicits = determine_implicit_constructors(explicit_items_found);
-            implicit_constructors_needed.insert(name.clone(), implicits);
+            let bases_items_found: Vec<_> = bases.iter().map_while(get_items_found).collect();
+            let field_types_items_found: Vec<_> =
+                field_types.iter().map_while(get_items_found).collect();
+            let has_rvalue_reference_fields = details.has_rvalue_reference_fields;
+
+            // Check that all the bases and field types are known first. This combined with
+            // iterating via [`depth_first`] means we can safely search in `items_found` for all of
+            // them.
+            //
+            // Conservatively, we will not acknowledge the existence of defaulted special member
```

Comment:
Hmmm, or maybe this part needs to change.

I don't see how to get examples/s2 working with this logic. There's a member variable which is a typedef to a template, so without analyzing templates or editing the source C++ I don't see how to teach this code that the type has a destructor.

But switching to assume that the special members do exist means we'll generate C++ code that doesn't compile sometimes, with no way to fix it without editing the source C++. That is the status quo, but I was hoping to improve upon it more. The rest of this PR still reduces the situations where we generate non-compiling C++.

Reply:
I don't have an immediate answer here.

Here are some pointers to relevant code just in case it helps.

The typedef part _might_ be resolvable using `type_converter::resolve_typedef`, but at best you'll be left with a templated type. `type_converter::get_templated_typename` synthesizes concrete type names for these things.

It's possible that `autocxx_bindgen` _does_ give enough information about templated types for us to make reasonable decisions here, though I doubt it: #349 is the can of worms related to templated types and I don't want to set off down that road until we absolutely have to. Right now the policy is pretty much "templated type? - nope!" and I'd prefer to stick with that as long as possible.

Wherever possible, `autocxx` tries to reject troublesome code by spotting such situations and generating an `Api::IgnoredItem` to give diagnostics to the user. See `error_reporter.rs`. That would be great. But more generally I'd be keen on any solution where we _detect_ this sort of troublesome situation and then prevent bad downstream code generation from occurring. For example maybe we record a Boolean to say "we're not sure whether this type has a destructor" and then condition downstream steps on it.

I'm sure you've thought all this through already: unfortunately I'm not going to have a chance to go into it in more depth for a few days.

Reply:
> Wherever possible, `autocxx` tries to reject troublesome code by spotting such situations and generating an `Api::IgnoredItem` to give diagnostics to the user. See `error_reporter.rs`. That would be great. But more generally I'd be keen on any solution where we _detect_ this sort of troublesome situation and then prevent bad downstream code generation from occurring. For example maybe we record a Boolean to say "we're not sure whether this type has a destructor" and then condition downstream steps on it.

This is effectively what the current version does (avoids generating downstream code that might be bad, but assuming it has no destructor). There's no `IgnoredItem` because I'm not sure where it should go in the generated Rust code for trait methods, and it's an implicit destructor anyways so there isn't anything to tie it to in the C++ source either.

Reply:
Actually, thinking about it, there's only one thing that we need to assume the templated type has for the s2 example: an accessible non-deleted destructor. That's pretty common, so probably a reasonable assumption to make. It's not true for my use case with `flatbuffers::Table`, but that's not templated so I can just `generate!` it. I'm going to run with that and see how far I get.

Reply:
Yes, fixing pointer/reference members got all the existing tests to pass, and assuming destructors exist got the s2 example to pass. This shouldn't be a problem any more.